### DISCUSSION TITLE
Entity remove - taking too long

### DISCUSSION THREAD

**Discussion by ElAndree** [Minestom/Minestom#669] [Help]
Hey there,

i am currently experiencing some troubles with falling entities again.
If i  call entity#remove, it takes sometimes a long time before the entity gets removed.
I tried to do a demonstrating video: https://youtu.be/vXvyETXlmAA

@TheMode can you maybe have a look at this?

Thank you!

Kind regards
Andre

```java
public class GravityBlockRule extends BlockPlacementRule {

    private long currentTime = -1;

    private ItemStack DROP_ITEM = ItemStack.of(getBlock().registry().material());

    public GravityBlockRule(@NotNull Block block) {
        super(block);
    }

    @Override
    public @NotNull Block blockUpdate(@NotNull Instance instance, @NotNull Point blockPosition, @NotNull Block currentBlock) {
        if (this.checkGravity(instance, blockPosition)) {
            instance.scheduler().buildTask(() -> instance.setBlock(blockPosition, Block.AIR)).delay(TaskSchedule.tick(2)).schedule();
            return currentBlock;
        }

        return currentBlock;
    }

    public boolean checkGravity(Instance instance, Point position) {
        Block below = instance.getBlock(position.blockX(), position.blockY() - 1, position.blockZ());

        if (!below.isAir() && !below.isLiquid()) {
            return false;
        }

        Pos initialPosition = new Pos(position.x() + 0.5f, Math.round(position.y()), position.z() + 0.5f);
        Entity fallingBlockEntity = new Entity(EntityType.FALLING_BLOCK);
        ((FallingBlockMeta) fallingBlockEntity.getEntityMeta()).setBlock(getBlock());
        fallingBlockEntity.setInstance(instance, initialPosition);
        fallingBlockEntity.scheduleNextTick(this::checkIfStillFalling);

        return true;
    }

    public void checkIfStillFalling(Entity entity) {
        if(currentTime > 0) {
            System.out.println("time: " + (System.currentTimeMillis() - currentTime));
        }

        currentTime = System.currentTimeMillis();

        Block below = entity.getInstance().getBlock((int) entity.getPosition().x(), (int) Math.round(entity.getPosition().y()) - 1, (int) entity.getPosition().z());


        if (BlockTags.DROP_ON_FALLING_BLOCK.contains(below.id())) {
            BlockHelper.dropItem(DROP_ITEM, entity.getPosition());
            entity.remove();
            return;
        }

        if (!below.isAir() && !below.isLiquid() && entity.getVelocity().isZero()) {
            entity.getInstance().setBlock(entity.getPosition(), getBlock());
            entity.remove();
            return;
        }

        entity.scheduleNextTick(this::checkIfStillFalling);
    }
}

```

**Comment by TheMode** [comment]
I'd recommend writing a more minimal example, as the code is not complete and hard to debug by hand.

`instance.scheduler().buildTask(() -> instance.setBlock(blockPosition, Block.AIR)).delay(TaskSchedule.tick(2)).schedule();` Why is this delay required?

`(int) Math.round(entity.getPosition().y())` can be replaced to `Pos#blockY()`, or even better, the whole line to `Block below = entity.getInstance().getBlock(entity.getPosition().sub(0,1,0);`

I will additionally mention that packets are flushed every 50ms (depends on TPS), which may add some latency (far from what is seen in your video)

**Reply by ElAndree** [reply]
Hey,

the delay is required, because else, you'll get problems in this case when you break the block with the "x" and all the blocks collapse into one block.
![image](https://user-images.githubusercontent.com/39100879/153717989-facdc832-9d87-4e12-9bc4-bd6d32456996.png)

blockY() can't be used here, because in the following case (lets say iron block = sand), it will drop / detect it instantly because blockY() says it has the y above the torch even when it just a minimal bit moving in the block below with the gravity, so i need to use the rawY and round it up/down.
Its hard to explain, i hope its understable.
![image](https://user-images.githubusercontent.com/39100879/153718089-3e8b02bb-ca94-48b8-a927-ea863f4d9ad4.png)

Could you clarify what you need for me as more minimal example?
Its reproducable with a minestom + this registered block rule.

Thank you in advance!

**Reply by ElAndree** [reply]
I found out one more piece of information - if it helps - the longer the sand has been falling, the longer it takes for the entity to be removed when calling the remove function.

**Reply by ElAndree** [reply]
Hey @TheMode i've written a small complete example reproducer (https://github.com/ElAndree/minestom-reproducer)
Could you maybe take a look at this?
I've experienced - if it helps - that the longer the falling block has been falling, the longer it takes for the entity to be removed when calling the remove function, but i dont have any clue why this is so.

Thank you in advance!

**Reply by TheMode** [reply]
What I meant by minimal reproducer is that most of your code is very likely unrelated to the issue, this could be as simple as spawning a falling block very high, remove the entity after a few time, and send a message to the player to have a rough understanding of the delay.

Right now you are asking me to debug your whole code.
